TAServers/lest
# LEST-94 - Add public API to lest package

## Ticket



Resolves [LEST-94](https://taservers.atlassian.net/browse/LEST-94)

## Changes

- Adds an easy to use object-oriented wrapper around the lest process
- Replaces the package "binary" with a new dedicated `/bin/lest.ts` binary that uses the new wrapper
- Fixes the binary not exiting with a failure code when the tests fail

## Impact

This unblocks a sizable chunk of work, specifically around tooling:

- Replacing the old composite action with a node-based action, providing additional functionality and fixing the issues with the current action
- Lays the groundwork for the VSCode extension with an extendable events API which will process a JSON reporter's output before passing to more granular `LestEvent`s

This also fixes a **major bug** with the current package:

- Currently, the package's binary will **always exit with a success code**, which makes it essentially unusable when running in automated pipelines. These changes add handling for the success state of lest, exposing it in an asynchronous way

## Testing

1. Create a new project which has Lest as a dependency
2. Add `lest.config.lua` and a test file with a failing test
3. Run `npx lest`
4. See that the process exits with a non-zero exit code
5. Create a JS/TS source file
6. Add `import { LestProcess } from "@taservers/lest";`
7. Add `console.log(await (new LestProcess()).run())`, replacing `await` with `then` if convenient
8. Run the source file
9. See `false` logged
10. Update the lua test to pass
11. Run `npx lest`
12. See that the process exits with a zero exit code
13. Run the source file
14. See `true` logged

## Helpful Links

[API Docs](https://taservers.github.io/lest/docs/api/expect)

[Project Board](https://taservers.atlassian.net/jira/software/c/projects/LEST/boards/4)

[Discord](https://discord.tasevers.com)


[LEST-94]: https://taservers.atlassian.net/browse/LEST-94?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
